src/main/java/com/google/opensesame/github/GitHubGetter.java

```diff
@@ -0,0 +1,54 @@
+package com.google.opensesame.github;
+
+import java.io.IOException;
+import org.kohsuke.github.GitHub;
+import org.kohsuke.github.GitHubBuilder;
+
+public final class GitHubGetter {
+  private static GitHub gitHub = null;
+
+  /**
+   * Gets the GitHub API interface or builds one if there is not one available.
+   *
+   * @return Returns the GitHub API interface.
+   * @throws IOException Throws IOException if there is an error building the GitHub interface.
+   */
+  public static GitHub getGitHub() throws IOException {
+    if (gitHub == null) {
+      buildGitHubInterface();
+    }
+
+    return gitHub;
+  }
+
+  /**
+   * Builds the GitHub interface from environment variables.
+   *
+   * <p>With AppEngine, the environment variables are set in the appengine-web.xml file.
+   *
+   * @throws IOException Throws IOException if there is an error building the GitHub interface.
+   */
+  private static void buildGitHubInterface() throws IOException {
+    String clientID = System.getenv("GITHUB_CLIENT_ID");
+    String clientSecret = System.getenv("GITHUB_CLIENT_SECRET");
+
+    if (clientID == null || clientID.isEmpty() || clientSecret == null || clientSecret.isEmpty()) {
+      System.err.println(
+          "GitHub authorization is not set. Please refer to the project README to"
+              + " configure this. Using unauthorized GitHub API for now.");
+      gitHub = GitHub.connectAnonymously();
+      return;
+    }
+
+    gitHub = new GitHubBuilder().withPassword(clientID, clientSecret).build();
+
+    try {
+      gitHub.checkApiUrlValidity();
+    } catch (IOException e) {
+      System.err.println(
+          "Invalid GitHub credentials. Please refer to the project README to"
+              + " configure this. Using unauthorized GitHub API for now.");
+      gitHub = GitHub.connectAnonymously();
```

Comment:
Setting `gitHub` in getGitHub makes it clearer that `buildGitHubInterface` is only meant to be called once and that the result is cached in gitHub for future use.

```suggestion
  public static GitHub getGitHub() throws IOException {
    if (gitHub == null) {
      gitHub = buildGitHubInterface();
    }

    return gitHub;
  }

  /**
   * Builds the GitHub interface from environment variables.
   *
   * <p>With AppEngine, the environment variables are set in the appengine-web.xml file.
   *
   * @throws IOException Throws IOException if there is an error building the GitHub interface.
   */
  private static GitHub buildGitHubInterface() throws IOException {
    String clientID = System.getenv("GITHUB_CLIENT_ID");
    String clientSecret = System.getenv("GITHUB_CLIENT_SECRET");

    if (clientID == null || clientID.isEmpty() || clientSecret == null || clientSecret.isEmpty()) {
      System.err.println(
          "GitHub authorization is not set. Please refer to the project README to"
              + " configure this. Using unauthorized GitHub API for now.");
      return GitHub.connectAnonymously();
    }

    authenticatedGitHub = new GitHubBuilder().withPassword(clientID, clientSecret).build();

    try {
      gitHub.checkApiUrlValidity();
      return authenticatedGitHub;
    } catch (IOException e) {
      System.err.println(
          "Invalid GitHub credentials. Please refer to the project README to"
              + " configure this. Using unauthorized GitHub API for now.");
     return GitHub.connectAnonymously();
```

Reply:
Also, if possible test `buildGitHubInterface`. To make `buildGitHubInterface` easier to test you might want to make the `System.getenv` calls in getGitHub and then pass the results as arguments to `buildGitHubInterface`.

Reply:
Totally agree with both comments. Thanks! Will add some testing to this.